Add lowLatency flag to CanvasRenderingContext2DSettings

yellowdoge opened the pull request:
This is a duplicate of the current state of @kenrussell's #4234 PR (that partially addresses #4087), so I can take it over from here.

Note that this PR is blocking KhronosGroup/WebGL#2753.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/4360/canvas.html" title="Last updated on Mar 20, 2019, 2:31 PM UTC (a2ae3f8)">/canvas.html</a>  ( <a href="https://whatpr.org/html/4360/c8f0426...a2ae3f8/canvas.html" title="Last updated on Mar 20, 2019, 2:31 PM UTC (a2ae3f8)">diff</a> )
<a href="https://whatpr.org/html/4360/references.html" title="Last updated on Mar 20, 2019, 2:31 PM UTC (a2ae3f8)">/references.html</a>  ( <a href="https://whatpr.org/html/4360/c8f0426...a2ae3f8/references.html" title="Last updated on Mar 20, 2019, 2:31 PM UTC (a2ae3f8)">diff</a> )

yellowdoge commented:
@annevk, @kenrussell, @jrmuizel, @grorg, @travisleithead, @fserb 

Gathering a wee bit the status:
(1) - Low Latency as a feature comprises two parts: one is decoupling the input event processing and rendering from the browser main loop, and another is rendering onto the front buffer. (Chrome ToT (M74) has the first part in all platforms, whereas currently only implements front buffer rendering on ChromeOS.)
(2) - The App can use either or both of the features described in (1) to decrease latency from input events to rendering, but of course at the expense of having -maybe- tearing of the rendered contents; this PR is about offering this possibility to application developers which anyway would need to be aware of the implications of what they are doing (see e.g. [this article](https://www.imgtec.com/blog/reducing-latency-in-vr-by-using-single-buffered-strip-rendering/)).

With that I'd say the best flag name might be `un`/`synchronized` (`synchronized: false` or `unsynchronized: true`), conveying the two parts of the desynchronization as described in (1) and doesn't look like something everyone would like to have in their apps. WDYT? (Feel free :+1: this message).

Of course the PR should reflect this tradeoff, if at least in some informatiional non-normative note.

kenrussell commented:
It's all personal preference, but I like `unsynchronized:true`, `desynchronized:true`, and `allowTearing:true`.

domenic commented:
I'll just note that boolean dictionary values [should default to false](https://heycam.github.io/webidl/#idl-dictionaries), so all of @kenrussell's latest suggestions are good in that regard (as is the OP `lowLatency: true`). But e.g. `synchronized: false` would be bad.

(Yes, `alpha: false` breaks this rule. That is a sad historical mistake.)

yellowdoge commented:
`unsynchronized: true` ping:  @annevk, @jrmuizel, @grorg, @travisleithead, @fserb, WDYT?

jdashg commented:
I generally do prefer `foo: true` over `unfoo: false`, so that `!unfoo` would just be `foo`.
If it's a strong recommendation that defaults should be false, I'd prefer `desynchronized`, which I think better reflects that it's allowing opting out of (the default) synchronization. (`unsynchronized` is acceptable, but not my preference)

yellowdoge commented:
So, `desynchronized: true` seems to be leading the pack, @annevk, @travisleithead, does it sounds good to you too?

othermaciej commented:
@yellowdoge I wrote in the previous PR for this that I don't really understand what this does, and I feel a bit more confused after your explanation. Some specific questions that came to mind:

- Why would a canvas context flag affect processing of input events? Does setting this flag on any context affect all input events? Also, I didn't see anything in the spec change about processing of input events, did I miss it? Or is this a separate feature that you described solely for context.

- I'm still not totally clear on what "rendering from the front buffer" means even after reading the Wikipedia reference on Multiple Buffering. 

- It doesn't seem like the spec has enough info to interoperable implement this, or to write test cases. Maybe that's ok for a performance hint?

- Is this feature intended to be implementable on and/or relevant to platforms other than Chrome OS?

annevk commented:
The name seems fine to me, but I'm a bit surprised @othermaciej's feedback on the prior PR went unaddressed. I'll try to be more careful ensuring all loose ends are tied up when we switch the place of discussion.

yellowdoge commented:
> @othermaciej's feedback on the prior PR went unaddressed.

I thought it better to agree on a name for the new functionality before even getting to writing the spec patch, since the former will (greatly) affect the latter. 

I can reply here to two of the concrete comments from https://github.com/whatwg/html/pull/4360#issuecomment-466260335:
> * Why would a canvas context flag affect processing of input events?

Why not? If we want to decrease latency between input event and pixels on screen, this is one of the sources of delay to cut on. Moreover, some elements (like <video>) produce output in a manner essentially decoupled from the event processing loop, so this operational mode would be nothing new to the platform.

>* Is this feature intended to be implementable on and/or relevant to platforms other than Chrome OS?

Yes, in my coarse description in https://github.com/whatwg/html/pull/4360#issuecomment-462931130 , there's nothing ChromeOS specific, and as a matter of fact we're exploring the implementation on Windows.

yellowdoge commented:
Folks I've uploaded a new version of the patch where hopefully things are clearer. I'm using the term `desynchronized` since it seems more consensual and nobody has expressed opinions against. I tried to explain in a note what is `front buffer` rendering, starting from the assumption that readers understand what is rendering, drawing, etc, pretty basic concepts IMHO but please give some feedback from spec-land as to what's the appropriate prerequisites of the reader. Note that `front-buffer` is just one of the ways to reduce latency (albeit a powerful one), but there's at least another one that is referenced.

othermaciej commented:
@travisleithead 
> I was similarly confused, but after @yellowdoge's answer, I understand now that there's no feature work to directly impact how input events are dispatched and processed[...]

The new spec diff says this (emphasis added)
> When a CanvasRenderingContext2D  object has its  desynchronized  flag set to true, then the user agent should optimize the rendering of the canvas to reduce the latency, as measured from input events to rasterization, by either **decoupling the input event processing from the normal event processing cycle**, bypassing the ordinary user agent/OS rendering algorithm, or both. Insofar as this mode involves bypassing the usual processing mechanisms of events or rasterization or both, it might introduce visible tearing artifacts. 

This totally sounds like it would affect input event processing. It describes this as a separate change from the rendering change that is also controlled by the flag. (And it doesn't have enough info to interoperably implement this alternate event processing, if that is what it's supposed to mean).

@yellowdoge can you clarify? Is this flag meant to affect event processing, and not just rendering?

yellowdoge commented:
> @yellowdoge can you clarify? Is this flag meant to affect event processing, and not just rendering?

Yes, as it says : 
> by either decoupling the input event processing from the normal event processing cycle, bypassing the ordinary user agent/OS rendering algorithm, or both. 

Reducing the latency can be achieved by reducing any of its components, namely: (a) input event delivery, (b) app logic processing, (c) drawing of the canvas (== execution of the drawing commands), and (d) OS rendering. If _any_ of them is reduced, then the overall latency is reduced, right? Depending on the UA and OS capabilities, reduction of the latency can be done in several ways: reducing (a), reducing (c), reducing (d) _or any combination of those_.

The flag is a hint -- but if the UA is absolutely unable to improve (== reduce) he latency as defined before, then `desynchronized` will read back false, meaning it's not supported.

jdashg commented:
Can we drop the 'changes input event processing' part? It seems like the primary adjustment we're making here is to desynchronize from the normal DOM update presentation atomicity.

Is there a need to go further than that right now, or can we move forward with just desynchronization from DOM update presentation atomicity?

othermaciej commented:
I tend to agree with @jdashg . Changing input event processing is pretty concerning here, and it seems better left out. Reasoning:

- It is surprising for a flag on a canvas graphics context object to change anything about input event handling.
- It's unclear whether the effect is global to all input events or only about events directed to the relevant canvas element somehow.
- While the rendering effects of this patch are not directly observable to web content (they only affect what the user sees on the screen), the event effects sound like they are. In particular, it sounds like timing of input events relative to other tasks or micro tasks could change. So there is interoperability risk if the event handling changes are implemented in slightly different ways.
- At the same time, the specific possible changes to event handling are not described here at all, so it's impossible to implement them interoperably based on the spec, or make test cases that reflect the spec.

An alternative would be to specify what the alternate event handling behavior is, in a way that is precise enough to make test cases (even if it is a MAY-level requirement).

othermaciej commented:
Further question about the term "front buffer rendering": could this also be accurately described as "single buffering" or "bypassing double/triple buffering"? I tried to look up resources on it and it sounds the same, but I might be missing a subtle difference. I ask because "single buffering" seems to be a more common term than "front buffer rendering", at least based on search engine frequency.

yellowdoge commented:
Guys PTAL:
- Removed the reference to the `input event`s and left it as `...desynchronizing the canvas paint cycle from the normal DOM cycle`, which is something we can control from the Canvas Context.  If this is still unclear or off the point, please suggest a wording on the diff (@jdashg, @travisleithead, @othermaciej).
- Added the reference to `multiple buffering`:
> A popular technique for reducing latency is called front buffer rendering, also known as single buffer rendering, where rendering happens in parallel and racily with the scanning out process.

annevk commented:
Aside: [please don't use "guys"](https://notapattern.net/2014/10/14/ways-men-in-tech-are-unintentionally-sexist/).

yellowdoge commented on source:
```diff
@@ -60390,20 +60390,19 @@ interface <dfn>Path2D</dfn> {
+  agent should optimize the rendering of the canvas to reduce the latency, as measured from
+  input events to rasterization, by either desynchronizing the canvas paint cycle from the normal
+  DOM cycle, bypassing the ordinary user agent/OS rendering algorithm, or both. Insofar as this
+  mode involves bypassing the usual processing mechanisms of events or rasterization or both, it
```
Done!

yellowdoge commented:
@travisleithead changes taken in, thanks!

@annevk: good to know, I thought "guys" was a generic, gender-neutral way to address a plurality.

@jdashg, @othermaciej: does the current patch LGTY ?

jdashg commented on source:
```diff
@@ -60379,11 +60384,44 @@ interface <dfn>Path2D</dfn> {
+  false. Initially, when the context is created, its <span
+  data-x="concept-canvas-desynchronized">desynchronized</span> flag must be set to false. When a
+  <code>CanvasRenderingContext2D</code> object has its <span
+  data-x="concept-canvas-desynchronized">desynchronized</span> flag set to true, then the user
```
Woah, we're allowing toggling desynchronized? I thought this was an immutable context creation option.

What's the rationale for toggling this in particular?

yellowdoge replied:
Not at all, IIRC this wording just mimics the other:
> The CanvasRenderingContext2D object also has an alpha flag, which can be set to true or false. Initially, when the context is created, its alpha flag must be set to true. When a CanvasRenderingContext2D object has its alpha flag set to false...

the only way to pass a value of either `alpha` or `desynchronized` is upon construction and since `getContext()` is idempotent, there's no way to change this value after construction.  I can rewrite the text here to make it more clear ?

yellowdoge replied:
@jdashg WDYT? (also @annevk what'd be the best way to encode that `desynchronized` value is latched upon creation and never modified?)

annevk replied:
Maybe "When a Canvas... object's alpha flag is false, the contents ..." to clarify there's not really a set operation for it.

yellowdoge replied:
Nuked the part where it said `Initially, when the context is created, its desynchronized flag must be set to false.` and changed the wording of the `true` part to `When an object's desynchronized flag is true...`, clarifying that there's no flag toggling possible.

jdashg replied:
Thanks!

yellowdoge commented:
@jdashg PTAnotherLook
 
@othermaciej: ping PTAL

(If the `alpha` value description sounds strange I can clean it up in another patch).

annevk commented:
I ended up making a whole bunch of minor changes to make alpha and desynchronized more similar while avoiding the problems folks had with their descriptions. I also spelled "back buffer" without hyphen as it seemed inconsistent with "front buffer". It probably warrants one more round of review though and I guess it'd be good if @othermaciej explicitly confirms he's okay with the approach now.

othermaciej commented:
I'll review the latest version this evening (Pacific time). I appreciate all the updates to address feedback.

othermaciej commented on source:
```diff
@@ -60316,36 +60317,38 @@ interface <dfn>Path2D</dfn> {
 
+    <p>If the <code
+    data-x="dom-CanvasRenderingContext2DSettings-desynchronized">desynchronized</code> member is
+    true, then the context can be <span
```
Should this say "may be" instead of "can be"?

yellowdoge replied:
Done.

yellowdoge replied:
Ah, sorry I uploaded the next PS before running the build and the linter failed:

> RFC2119 keyword in domintro (use: might, can, has to, or override with <!--non-normative-->must):
> 60330:     true, then the context may be <span

So I substituted `may` with `might`.

othermaciej commented on source:
```diff
@@ -60376,14 +60378,43 @@ interface <dfn>Path2D</dfn> {
+  <p>The <code>CanvasRenderingContext2D</code> object also has a <dfn
+  data-x="concept-canvas-desynchronized">desynchronized</dfn> boolean. When a
+  <code>CanvasRenderingContext2D</code> object's <span
+  data-x="concept-canvas-desynchronized">desynchronized</span> is true, then the user agent should
```
This ought to be a may-level requirement rather than should-level. On many platforms, it may not be necessary or helpful (or may not be possible) to do anything different than normal rendering.

yellowdoge replied:
Done.

othermaciej commented on source:
```diff
@@ -60376,14 +60378,43 @@ interface <dfn>Path2D</dfn> {
+  spec=MULTIPLEBUFFERING></p>
+
+  <p class="note">The <span data-x="concept-canvas-desynchronized">desynchronized</span> boolean is
+  useful and necessary when implementing certain kinds of applications, such as drawing
```
Although this is a non-normative note, it's not clear to me that it's necessary. Drawing with an Apple Pencil on a on an iPad Pro (whether on an HTML 2D canvas or not) is just fine without doing single-buffering or any other special case optimization that hurts quality. So I'd suggest replacing the words "is useful and necessary" with "may be useful".

annevk replied:
Since it's a note "can be useful". (I agree with your suggestion of using "may" earlier though, as it's normative there.)

yellowdoge replied:
Done.

yellowdoge commented:
@othermaciej's comments taken in in fd0967f, all concerns should be addressed now!

yellowdoge commented:
@annevk WDYT?

annevk commented on source:
```diff
@@ -60376,14 +60378,43 @@ interface <dfn>Path2D</dfn> {
+  <code>CanvasRenderingContext2D</code> object's <span
+  data-x="concept-canvas-desynchronized">desynchronized</span> is true, then the user agent may
+  optimize the rendering of the canvas to reduce the latency, as measured from input events to
+  rasterization, by desynchronizing the canvas paint cycle from the normal DOM cycle, bypassing the
```
"normal DOM cycle" should be "event loop", right?

yellowdoge replied:
Yes ofc, done.

yellowdoge commented:
Buglist:
- Chrome: https://crbug.com/788439
- FF: https://bugzilla.mozilla.org/show_bug.cgi?id=1536809
- WebKit: https://bugs.webkit.org/show_bug.cgi?id=196002

yellowdoge commented:
@annevk changes taken in; when you say "question" in https://github.com/whatwg/html/pull/4360#pullrequestreview-216591833 , do you refer to the change `s/normal DOM cycle/event loop/` or something else ? (Went back in the conversation history but didn't see any other questions open).

annevk commented:
That was it, thanks!

annevk commented:
Should landing this close #4087?

yellowdoge commented:
@annevk https://github.com/whatwg/html/pull/4360#issuecomment-474867305 : this PR should close the 2D context part, there's a WebGL PR companion waiting on it somewhere (@kenrussell )

othermaciej commented:
@annevk I'm not sure it's quite right right that this can't be tested. While the effect on the graphics system is not testable, I think it's possible to write a test that checks that the `desynchronized` flag is correctly round-tripped when set via `getContext()` and retrieved via `getContextAttributes()`.  It's probably worth it to make a test like that, since it's a requirement even on implementations that don't do anything special for `desynchronized`.

annevk commented:
Thanks @othermaciej, you're right. @yellowdoge will you add such a test? It's okay if it's done via Chromium's commit process and automatically upstreamed.

yellowdoge commented:
Yeah in all likelihood I'll add that test and will be upstreamed.

annevk commented:
Okay, I'll mention https://bugs.chromium.org/p/chromium/issues/detail?id=788439 in the commit message as a place to look for tests then. We try to land them roughly at the same time, but this is minor enough that I'm okay with some delay.

annevk commented:
So I thought that since the initial commit was authored by @kenrussell, a squash & merge combined with [co-authored-by](https://help.github.com/en/articles/creating-a-commit-with-multiple-authors) for @yellowdoge would attribute the right folks on the commit, but somehow only @yellowdoge gets acknowledged. I guess the mistake I made is that if you use co-authored-by you need a line for each author? Sorry about that.

kenrussell commented:
Thanks but it's not necessary to cite me - @yellowdoge rewrote everything.
